File: CIS343/prj1/Minesweeper.c
/************************************************************************
 * minesweeper.c														*
 *																		*
 * Author(s): Isaac Smith
 *            Sam Essenburg
 ***********************************************************************/

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>




#define BOARD_SIZE_MIN 5
#define BOARD_SIZE_MAX 15
#define PCT_MINES_MIN 10
#define PCT_MINES_MAX 70


typedef struct {
	bool is_mine;
	int mines;
	bool visible;
}Cell;

typedef enum {WON, LOST, INPROGRESS} Status;

/************************************************************************
 * YOU MUST NOT DEFINE ANY GLOBAL VARIABLES (i.e., OUTSIDE FUNCTIONS).  *
 * COMMUNICATION BETWEEN FUNCTIONS MUST HAPPEN ONLY VIA PARAMETERS.     *                           *
 ************************************************************************/

/************************************************************************
 * Function declarations/prototypes										*
 ************************************************************************/
void displayMenu();

int getBoardSize();

int getPercentMines();

void initBoard(int size, Cell board[][size]);

void placeMinesOnBoard(int size, Cell board[][size], int nbrMines);

void fillInMineCountForNonMineCells(int size, Cell board[][size]);

int nbrOfMines(int size, Cell board[][size]);

int getNbrNeighborMines(int row, int col, int size, Cell board[][size]);

void displayBoard(int size, Cell board[][size], bool displayMines);

Status selectCell(int row, int col, int size, Cell board[][size]);

int nbrVisibleCells(int size, Cell board[][size]);

void setImmediateNeighborCellsVisible(int row, int col, int size, Cell board[][size]);

void setAllNeighborCellsVisible(int row, int col, int size, Cell board[][size]);


/************************************************************************
 * Main driver of the program. Uses the functions defined above.		*
 ************************************************************************/
int main(){
	int row, col, size, nbrMines;
	char command;
	bool displayMines = false;
	Status gameState = INPROGRESS;

	printf("!!!!!WELCOME TO THE MINESWEEPER GAME!!!!!\n\n");
	size = getBoardSize();
	

	// declare 2D array of cells
	Cell board[size][size];
	initBoard(size,board);

	// determine number of mine to place on the board
	nbrMines = (int) (size * size * (getPercentMines()/100.0));
	printf("Num of Mines: %d\n", nbrMines);

	// place mines randomly on the board
	placeMinesOnBoard(size,board,nbrMines);

	// For each non-mine cell, sets the neighboring mine count
	fillInMineCountForNonMineCells(size,board);

	displayBoard(size,board,displayMines);

	while (true) {
		printf("Enter command (m/M for command menu): ");
		scanf("%c",&command);
		getchar();

		switch (command) {
			case 'm': case 'M':
				displayMenu();
				break;

			case 'c': case 'C':
				do {
					printf("Enter row and column of cell: ");
					scanf("%d%d",&row,&col);
					getchar();
					if (row < 1 || row > size || col < 1 || col > size) {
						printf("Invalid row or column values. Try again.\n");
					}
				} while (row < 1 || row > size || col < 1 || col > size);
				row--;
				col--;
				gameState = selectCell(row,col,size,board);
				displayBoard(size,board,displayMines);
				break;

			case 's': case 'S':
				displayMines = true;
				displayBoard(size,board,displayMines);
				break;

			case 'h': case 'H':
				displayMines = false;
				displayBoard(size,board,displayMines);
				break;

			case 'b': case 'B':
				displayBoard(size,board,displayMines);
				break;

			case 'q': case 'Q':
				printf("Bye.\n");
				return 0;

			default:
				printf("Invalid command. Try again.\n");
		}

		// display appropriate message if the game is over
		if (gameState == WON) {
			printf("You found all the mines. Congratulations. Bye.");
			return 0;
		} else if (gameState == LOST) {
			printf("Oops. Sorry, you landed on a mine. Bye");
			return 0;
		}
	}

	return 0;
}

/************************************************************************
 * Displays list of commands allowed at the prompt						*
 ************************************************************************/
void displayMenu()
{
  printf("List of available commands:\n");
  printf("   Show Mines: s/S\n");
  printf("   Hide Mines: h/H\n");
  printf("   Select Cell: c/C\n");
  printf("   Display Board: b/B\n");
  printf("   Display Menu: m/M\n");
  printf("   Quit: q/Q\n\n");
}

/************************************************************************
 * Initializes the fields of each cell on the board as follows:			*
 * 		is_mine field to false											*
 * 		mines field to 0												*
 * 		visible field to false											*
 ************************************************************************/
void initBoard(int size, Cell board[][size])
{
  // TO DO

  for(int row = 0; row < size; row++){
    for(int col = 0; col < size; col++){
      board[row][col].is_mine = false;
      board[row][col].visible = false;
      board[row][col].mines = 0;
    }
  }
	
}

/************************************************************************
 * Places the specified number of mines randomly on the board			*
 ************************************************************************/
void placeMinesOnBoard(int size, Cell board[][size], int nbrMines)
{
  // TO DO

  int i = 0;

  //printf("%d\n", nbrMines);
  while(i < nbrMines){
    srand(time(NULL));
    int row = rand()%size;
    int col = rand()%size;
    if(board[row][col].is_mine == false){
      board[row][col].is_mine = true;
      i = i + 1;
    }else{
      board[row][col].is_mine = true;
    }
  }
}

/************************************************************************
 * For each non-mine cell on the board, set the "mines" field to the	*
 * number of mines in the immediate neighborhood.						*
 ************************************************************************/
void fillInMineCountForNonMineCells(int size, Cell board[][size])
{
  // TO DO

  for(int row = 0; row < size; row++){
    for(int col = 0; col < size; col++){
      if(board[row][col].is_mine == false){
 //       for(int i = row-1; i < row + 2; i++){
  //        for(int j = col-1; j < col + 2; j++){
  //          if(board[i][j].is_mine == true && i < size && j < size)
  //            board[row][col].mines++;
            board[row][col].mines = getNbrNeighborMines(row, col, size, board);
          }
        }	
      }
    }
 

/************************************************************************
 * Counts and returns the number of mines on the board					*
 ************************************************************************/
int nbrOfMines(int size, Cell board[][size])
{
  int count = 0;

  for(int row = 0; row < size; row++){
    for(int col = 0; col < size; col++){
      if(board[row][col].is_mine == true)
        count++; 
    }
  }  

  return count;
}

/************************************************************************
 * Returns the number of mines in the immediate neighborhood of a cell	*
 * at location (row,col) on the board.									*
 ************************************************************************/
int getNbrNeighborMines(int row, int col, int size, Cell board[][size])
{
  int count = 0;

  if(row == 0 && col == 0){
    if(board[row][col+1].is_mine)
      count++;
    if(board[row+1][col].is_mine)
      count++;
    if(board[row+1][col+1].is_mine)
      count++;
  }
  else if(col == 0){
    for(int i = row -1; i <  row + 2;i++){
      for(int j = col; j < col + 2;j++){
        if(board[i][j].is_mine == true){
          count++;
        }
      }
    }
  }
  else if(row == 0){
    for(int i = row; i < row + 2; i++){
      for(int j = col - 1; j < col + 2; j++ ){
        if(board[i][j].is_mine == true)
          count++;
      }
    }
  }
  else{
  for(int i = row - 1; i < row + 2; i++){
    for(int j = col - 1; j < col + 2; j++){
      if(board[i][j].is_mine == true && i < size && j < size)
        count++;
    }
  }  
  }

  return count;
}

/************************************************************************
 * Displays the board. If a cell is not currently visible and has a		*
 * mine, show the mine if the displayMines is true. Used for debugging	*
 * and testing purposes.												*
 ************************************************************************/
void displayBoard(int size, Cell board[][size], bool displayMines)
{
  // TO DO
  printf(" ");

  for(int col = 0; col < size; col++){
    if(col < 9)
      printf("  %d", col+1);
    else
      printf(" %d", col+1);
  }
  printf("\n");
  
  for(int row = 0; row < size; row++){
    if(row < 9){
        printf("%d  ", row + 1);  
    }
    else
        printf("%d ", row + 1);
    for(int c = 0; c < size; c++){
      if(displayMines == true && board[row][c].is_mine){
        printf("*  ");
      }
      else if(board[row][c].visible == true){
        printf("%d  ", board[row][c].mines);
      }
      else{
        printf("?  ");
      }
    }
    printf("\n");
  }
	
	printf(" \n");
}

/************************************************************************
 * Prompts the user for board size, reads and validates the input		*
 * entered, and returns the integer if it is within valid range.		*
 * repeats this in a loop until the user enters a valid value.			*
 ************************************************************************/
int getBoardSize()
{
  int size = 0;
  printf("Please enter a board size. Must be >=5 and <=15\n\n");
  scanf("%d",&size);
	
  if(size < BOARD_SIZE_MIN){
    printf("sorry please enter a size >= 5\n\n");
    getBoardSize();
  }else if(size > BOARD_SIZE_MAX){
    printf("Sorry, please enter a size <= 15\n\n");
    getBoardSize();
  }
  
  return size;

}

/************************************************************************
 * Prompts the user for percentage of mines to place on the board,		*
 * reads and validates the input entered, and returns the integer if it	*
 * is within valid range. repeats this in a loop until the user enters	*
 * a valid value for board size.										*
 ************************************************************************/
int getPercentMines()
{
  int percent = 0;

  int valid = 0;
  // TO DO
  while(valid != 1){
    printf("Please choose the percentage of mines on the board.\n\n");
    scanf("%d",&percent);
    
    if(percent <= 0 || percent >=100){
      printf("Sorry the percent must be between 1 and 99. Please try again.\n\n");
    }else{
      valid = 1;
    }
  }
  return percent;
}

/************************************************************************
 * Process cell selection by user during the game						*
 ************************************************************************/
Status selectCell(int row, int col, int size, Cell board[][size])
{
  // TO DO
  int vis;
  if(board[row][col].is_mine){
    printf("Oop. Sorry, you landed on a mine. Bye");
    
    return LOST;
  }
  else{
    board[row][col].visible = true;
    setAllNeighborCellsVisible(row, col, size, board);
    for(int r = 0; r < size; r++){
      for(int c = 0; c < size; c++){
        if(board[r][c].visible == true && board[r][c].is_mine != true){
          vis++;
        }
      }
    }
    if(vis == ((size*size)-nbrOfMines(size,board))){
      return WON;
    }
  }

  return INPROGRESS;
}

/************************************************************************
 * Returns the number of cells that are currently visible.				*
 ************************************************************************/
int nbrVisibleCells(int size, Cell board[][size])
{
  int count = 0;

  for(int row = 0; row < size; row++){
    for(int col = 0; col < size; col++){
      if(board[row][col].visible == true)
        count++;
    }
  }
  
  return count;
}

/************************************************************************
 * If the mine count of a cell at location (row,col) is zero, then make	*
 * the cells ONLY in the immediate neighborhood visible.				*
 ************************************************************************/
void setImmediateNeighborCellsVisible(int row, int col, int size, Cell board[][size])
{
  if(board[row][col].mines == 0){
    for(int i = row - 1; i < row + 2; i++){
      for(int j = col - 1; j < col + 2; j++){
        board[i][j].visible = true;
      }
    }
  }
}

/************************************************************************
 * If the mine count of a cell at location (row,col) is zero, then make	*
 * the cells in the immediate neighborhood visible and repeat this		*
 * process for each of the cells in this set of cells that have a mine	*
 * count of zero, and so on.											*
 ************************************************************************/
void setAllNeighborCellsVisible(int row, int col, int size, Cell board[][size])
{

  
  if(board[row][col].mines == 0){
     for(int i = row - 1; i < row + 2; i++){
      for(int j = col - 1; j < col + 2; j++){
       board[i][j].visible = true;
         if(board[i][j].mines == 0){
          for(int r = i - 1; r < i + 2; r++){
            for(int c = j - 1; c < j + 2; c++){
              if(r > 0 && c > 0 && r < size-1 && c < size-1){
                board[r][c].visible == true;
              }
            }
          }
        }
      }
    }
  }
}

File: CIS343/prj5/Minesweeper.rb
#
#  Minesweeper game
#
#  Author(s): Isaac Smith and Sam Essenburg
#

#
# module that defines global constants
#
module Constants
  WON = 0
  LOST = 1
  INPROGRESS = 2
  BOARD_SIZE_MIN = 5
  BOARD_SIZE_MAX = 15
  PCT_MINES_MIN = 10
  PCT_MINES_MAX = 70
end

#
# Cell class represents a cell on the minesweeper board
#
class Cell
  
  attr_accessor :is_mine, :nbr_mines, :visible
  
  def initialize(nbr_mines,is_mine,visible)
    self.nbr_mines=(nbr_mines)
    self.is_mine=(is_mine)
    self.visible=(visible)
  end
  
  def to_s
    "#@is_mine #@nbr_mines #@visible"
  end
end

#
# Minesweeper class represents the game board and contains game logic
#
class Minesweeper
  
  # initializes a Minesweeper object
  def initialize(board_size,percent_mines)
    @board_size = board_size
    @nbr_mines = (board_size * board_size * (percent_mines/100.0)).to_i
    
    # setup a 2-dimensional array of Cell objects
    @board = Array.new(board_size)
    @board.fill { |i| Array.new(board_size) }
    for i in 0...board_size
      for j in 0...board_size
        @board[i][j] = Cell.new(0,false,false)
      end
    end
    
    place_mines_on_board()
    fill_in_minecount_for_non_mine_cells()
  end
  
  # places mines randomly on the board
  def place_mines_on_board
    count = 0
    while count < @nbr_mines
      x = Random.rand(@board_size) 
      y = Random.rand(@board_size)
      
      if !@board[x][y].is_mine
        @board[x][y].is_mine = true
        count += 1
      end
    end
  end
  
  # for each non-mine cell on the board, set @nbr_mines of each Cell on the board
  # to the number of mines in the immediate neighborhood.
  def fill_in_minecount_for_non_mine_cells
    for i in 0...@board_size
      for j in 0...@board_size
        if !@board[i][j].is_mine
          @board[i][j].nbr_mines = get_nbr_neighbor_mines(i,j)
        end
      end
    end
  end
    
  # processes cell selection by user during the game
  # returns Constants::WON, Constants::LOST, or Constants::INPROGRESS
  def select_cell(row,col)
    @board[row][col].visible = true
    
    if @board[row][col].is_mine
      return Constants::LOST
    end
    if @board[row][col].nbr_mines == 0
      set_all_neighbor_cells_visible(row,col)
    end
    if @nbr_mines + nbr_visible_cells == @board_size*@board_size
      return Constants::WON
    end
    
    return Constants::INPROGRESS
  end
  
  # returns the number of mines in the immediate neighborhood of a cell
  # at location (row,col) on the board.
  def get_nbr_neighbor_mines(row,col)
    count = 0
    
    	#check left, diagupleft, above, diagupright, right
	if col > 0 && @board[row][col-1].is_mine
  		count += 1
  	end
  	if (col > 0 && row > 0) && @board[row-1][col-1].is_mine
  		count += 1
  	end
  	if (row > 0 && @board[row-1][col].is_mine)
  		count += 1
  	end
  	if (row > 0 && col < @board_size-1) && @board[row-1][col+1].is_mine
  		count += 1
  	end
  	if col < @board_size-1 && @board[row][col+1].is_mine 
  		count += 1
  	end
  	#check diagdownright, below, diagdownleft
  	if (row < @board_size-1 && col < @board_size-1) && @board[row+1][col+1].is_mine
  		count += 1
  	end
  	if row < @board_size-1 && @board[row+1][col].is_mine
  		count += 1
  	end
  	if (row < @board_size-1 && col > 0) && @board[row+1][col-1].is_mine
  		count += 1
  	end
  	return count
  end
  
  # returns the number of cells that are currently visible on the board
  def nbr_visible_cells
    count = 0
    
    for i in 0...@board_size
      for j in 0...@board_size
        if @board[i][j].visible
          count += 1
        end
      end
    end
    return count
  end
  
  # if the mine count of a cell at location (row,col) is zero, then make 
  # the cells in the immediate neighborhood visible and repeat this
  # process for each of the cells in this set of cells that have a mine
  # count of zero, and so on.
  def set_all_neighbor_cells_visible(row,col)
    #check left
    if col > 0 && !@board[row][col-1].visible
    	select_cell(row, col-1)
    end
    #check diag up left
    if (col > 0 && row > 0) && @board[row-1][col-1].visible
    	select_cell(row-1, col-1)
    end
    #check above
    if (row > 0 && !@board[row-1][col].visible)
    	select_cell(row-1, col)
    end
    #check diag up right
    if ((row > 0 && col < @board_size-1) && !@board[row-1][col+1].visible)
    	select_cell(row-1, col+1)
    end
    #check right
    if (col < @board_size-1 && !@board[row][col+1].visible)
    	select_cell(row, col+1)
    end
    #check diag down right
    if ((row < @board_size-1 && col < @board_size-1) && !@board[row+1][col+1].visible)
    	select_cell(row+1, col+1)
    end
    #check below
    if (row < @board_size-1 && !@board[row+1][col].visible)
    	select_cell(row+1, col)
    end
    #check diag down left
    if ((row < @board_size-1 && col > 0) && !@board[row+1][col-1].visible)
    	select_cell(row+1, col-1)
    end
  end
  
  # returns a string representation of the board
  def to_s(display_mines=false)
    str = ""
    for i in 0...@board_size
      str << (i == 0 ? sprintf("%6d",i+1) : sprintf("%3d",i+1))
    end
    str << "\n"
    
    for i in 0...@board_size
      str << sprintf("%3d",i+1)
      for j in 0...@board_size
        if @board[i][j].visible
          str << (@board[i][j].is_mine ? sprintf("  *") : sprintf("%3d", @board[i][j].nbr_mines))
        else
          str << (display_mines && @board[i][j].is_mine ? sprintf("  *") : sprintf("  ?"))
        end
      end
      str << "\n"
    end
    
    str
  end
  
  # make these methods private
  private :place_mines_on_board, :fill_in_minecount_for_non_mine_cells,
        :get_nbr_neighbor_mines, :nbr_visible_cells, 
        :set_all_neighbor_cells_visible

end

#
# main method that provides the user interface to the game
#
def main
  display_mines = false
  game_state = Constants::INPROGRESS
  
  board_size = get_board_size()
  percent_mines = get_percent_mines()
  board = Minesweeper.new(board_size,percent_mines)
  puts board.to_s(display_mines)
  
  while true do
    print "Enter command (m/M for menu): "
    command = STDIN.gets.strip
    case command
      when 'm' || 'M'
        display_menu()
        
      when 'c' || 'C'
        row = 0
        col = 0
        while row < 1 || row > board_size || col < 1 || col > board_size do
          print "Enter row and column of cell: "
          tokens = STDIN.gets.strip.split(' ')
          row = tokens[0].to_i
          col = tokens[1].to_i
          if row < 1 || row > board_size || col < 1 || col > board_size
            puts "Invalid row or column values. Try again."
          end
        end
        game_state = board.select_cell(row-1,col-1)
        puts board.to_s(display_mines)
        
      when 's' || 'S'
        display_mines = true
        puts board.to_s(display_mines)
        
      when 'h' || 'H'
        display_mines = false
        puts board.to_s(display_mines)
        
      when 'b' || 'B'
        puts board.to_s(display_mines)
        
      when 'q' || 'Q'
        puts "Bye."
        return
        
      else
        puts "Invalid command. Try again."
    end # end case
    
    if game_state == Constants::WON
      puts "You found all the mines. Congratulations. Bye."
      return
    elsif game_state == Constants::LOST
      puts "Whoops. you landed on a mine. Bye"
      return
    end
  end   # end while
  
end

#
# Displays command menu
#
def display_menu
  puts "List of available commands:"
  puts "   Show Mines: s/S"
  puts "   Hide Mines: h/H"
  puts "   Select Cell: c/C"
  puts "   Display Board: b/B"
  puts "   Display Menu: m/M"
  puts "   Quit: q/Q"
end

#
# Prompts the user for board @board_size, reads and validates the input
# entered, and returns the integer if it is within valid range.
# repeats this in a loop until the user enters a valid value.
#
def get_board_size
  @board_size = 0
  while @board_size < Constants::BOARD_SIZE_MIN || @board_size > Constants::BOARD_SIZE_MAX do
    print "Enter board @board_size (#{Constants::BOARD_SIZE_MIN} .. #{Constants::BOARD_SIZE_MAX}): "
    @board_size = STDIN.gets.strip.to_i
    if (@board_size < Constants::BOARD_SIZE_MIN || @board_size > Constants::BOARD_SIZE_MAX) 
      puts "Invalid board @board_size. Try again."
    end
  end
  @board_size
end

#
# Prompts the user for percentage of mines to place on the board,
# reads and validates the input entered, and returns the integer if it
# is within valid range. repeats this in a loop until the user enters
# a valid value for board @board_size.
#
def get_percent_mines
  percent_mines = 0
  while percent_mines < Constants::PCT_MINES_MIN || percent_mines > Constants::PCT_MINES_MAX do
    print "Enter percentage of mines on the board (#{Constants::PCT_MINES_MIN} .. #{Constants::PCT_MINES_MAX}): "
    percent_mines = STDIN.gets.strip.to_i
    if (percent_mines < Constants::PCT_MINES_MIN || percent_mines > Constants::PCT_MINES_MAX)
      puts "Invalid value for percentage. Try again."
    end
  end
  percent_mines
end
 
#
# invoke main
#
main()
